src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -298,45 +313,74 @@ private void OnWriteCompleted(WriteContext writeContext)
             var status = writeContext.WriteStatus;
             var error = writeContext.WriteError;
 
-            if (error != null)
+            PoolWriteContext(writeContext);
+
+            if (error == null)
             {
-                _lastWriteError = new IOException(error.Message, error);
+                // _numBytesPreCompleted can temporarily go negative in the event there are
+                // completed writes that we haven't triggered callbacks for yet.
+                _numBytesPreCompleted -= bytesWritten;
 
-                // Abort the connection for any failed write.
-                _connection.Abort();
+                CompleteFinishedWrites(status);
             }
+            else if (error != null)
+            {
+                // Abort the connection for any failed write
+                // Queued on threadpool so get it in as first op.
+                _connection.Abort();
 
-            PoolWriteContext(writeContext);
-
-            // _numBytesPreCompleted can temporarily go negative in the event there are
-            // completed writes that we haven't triggered callbacks for yet.
-            _numBytesPreCompleted -= bytesWritten;
+                CancelOutstandingTasks(error);
+            }
+        }
 
+        private void CompleteFinishedWrites(int status)
+        {
             // bytesLeftToBuffer can be greater than _maxBytesPreCompleted
             // This allows large writes to complete once they've actually finished.
             var bytesLeftToBuffer = _maxBytesPreCompleted - _numBytesPreCompleted;
             while (_tasksPending.Count > 0 &&
-                   (int)(_tasksPending.Peek().Task.AsyncState) <= bytesLeftToBuffer)
+                   (_tasksPending.Peek().BytesToWrite) <= bytesLeftToBuffer)
             {
-                var tcs = _tasksPending.Dequeue();
-                var bytesToWrite = (int)tcs.Task.AsyncState;
+                var waitingTask = _tasksPending.Dequeue();
+                var bytesToWrite = waitingTask.BytesToWrite;
 
                 _numBytesPreCompleted += bytesToWrite;
                 bytesLeftToBuffer -= bytesToWrite;
 
-                if (_lastWriteError == null)
+                if (waitingTask.Token.IsCancellationRequested)
                 {
-                    _threadPool.Complete(tcs);
+                    _threadPool.Cancel(waitingTask.CompletionSource);
```

Comment:
At this point since the write has actually already completed, there isn't really any need to cancel the task even if the token has somehow tripped. We might as well complete them all as before.

In practice, once we start registering with cancellable tokens and unregistering prior to completing/canceling the TCSs, it's pretty unlikely we would get to this point with a canceled token.